File: kernel_design/executable_spec/assembly/assignment_dispatch_application.py
from __future__ import annotations

import logging
from collections.abc import Callable
from dataclasses import dataclass
from functools import partial
from threading import Event, Lock, Thread
from time import monotonic

from ..assignment_dispatch import (
    TaskItemDispatchConfig,
    TaskItemDispatchPacer,
    TaskRunningActivationConfig,
    TaskRunningActivationPacer,
    TaskWorkerAllocationConfig,
    TaskWorkerAllocationPacer,
)


_LOGGER = logging.getLogger(__name__)
_RoundOperation = Callable[[], int]


@dataclass(frozen=True)
class AssignmentDispatchApplicationConfig:
    worker_allocation: TaskWorkerAllocationConfig
    running_activation: TaskRunningActivationConfig
    task_item_dispatch: TaskItemDispatchConfig
    worker_allocation_interval_millis: int
    running_activation_interval_millis: int
    task_item_dispatch_interval_millis: int

    def __post_init__(self) -> None:
        if any(
            interval_millis <= 0
            for interval_millis in (
                self.worker_allocation_interval_millis,
                self.running_activation_interval_millis,
                self.task_item_dispatch_interval_millis,
            )
        ):
            raise ValueError("assignment-dispatch intervals must be positive")


class AssignmentDispatchApplication:
    """Application lifecycle for independent assignment-dispatch pacers."""

    def __init__(
        self,
        worker_allocation_pacer: TaskWorkerAllocationPacer,
        running_activation_pacer: TaskRunningActivationPacer,
        task_item_dispatch_pacer: TaskItemDispatchPacer,
    ) -> None:
        self.worker_allocation_pacer = worker_allocation_pacer
        self.running_activation_pacer = running_activation_pacer
        self.task_item_dispatch_pacer = task_item_dispatch_pacer
        self._lifecycle_lock = Lock()
        self._stop_event: Event | None = None
        self._threads: tuple[Thread, ...] = ()

    def start(self, *, config: AssignmentDispatchApplicationConfig) -> None:
        """Start one sequential background loop for each pacer."""
        with self._lifecycle_lock:
            if self._threads:
                raise RuntimeError("assignment-dispatch application is already started")

            stop_event = Event()
            loops = (
                (
                    "worker-allocation",
                    config.worker_allocation_interval_millis,
                    partial(
                        self.worker_allocation_pacer.allocate_candidate_workers,
                        config=config.worker_allocation,
                    ),
                ),
                (
                    "running-activation",
                    config.running_activation_interval_millis,
                    partial(
                        self.running_activation_pacer.activate_running_visible_tasks,
                        config=config.running_activation,
                    ),
                ),
                (
                    "task-item-dispatch",
                    config.task_item_dispatch_interval_millis,
                    partial(
                        self.task_item_dispatch_pacer.dispatch_task_items,
                        config=config.task_item_dispatch,
                    ),
                ),
            )
            threads = tuple(
                Thread(
                    name=f"assignment-dispatch-{loop_name}",
                    target=self._run_loop,
                    kwargs={
                        "loop_name": loop_name,
                        "stop_event": stop_event,
                        "interval_millis": interval_millis,
                        "operation": operation,
                    },
                    daemon=False,
                )
                for loop_name, interval_millis, operation in loops
            )

            started_threads: list[Thread] = []
            try:
                for thread in threads:
                    thread.start()
                    started_threads.append(thread)
            except Exception:
                stop_event.set()
                for started_thread in started_threads:
                    started_thread.join()
                raise

            self._stop_event = stop_event
            self._threads = threads

    def stop(self, *, timeout_millis: int = 5_000) -> None:
        """Stop all loops and wait for any in-flight bounded round."""
        if timeout_millis <= 0:
            raise ValueError("stop timeout must be positive")

        with self._lifecycle_lock:
            threads = self._threads
            stop_event = self._stop_event
            if not threads or stop_event is None:
                return
            stop_event.set()

        deadline = monotonic() + timeout_millis / 1_000
        for thread in threads:
            thread.join(timeout=max(0.0, deadline - monotonic()))

        alive_threads = tuple(thread.name for thread in threads if thread.is_alive())
        if alive_threads:
            raise TimeoutError(
                "assignment-dispatch loops did not stop: "
                + ", ".join(alive_threads)
            )

        with self._lifecycle_lock:
            if self._threads == threads:
                self._threads = ()
                self._stop_event = None

    @staticmethod
    def _run_loop(
        *,
        loop_name: str,
        stop_event: Event,
        interval_millis: int,
        operation: _RoundOperation,
    ) -> None:
        while not stop_event.is_set():
            try:
                operation()
            except Exception:
                _LOGGER.exception("assignment-dispatch %s round failed", loop_name)
            if stop_event.wait(interval_millis / 1_000):
                return

File: kernel_pacer_jvm/src/main/java/com/xa/mass/kernel/pacer/assignment/AssignmentDispatchApplication.java
package com.xa.mass.kernel.pacer;

import java.util.List;
import java.util.ArrayList;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

final class AssignmentDispatchApplication {

    private static final System.Logger LOGGER = System.getLogger(
            AssignmentDispatchApplication.class.getName()
    );
    private static final long START_ROLLBACK_TIMEOUT_MILLIS = 5_000;
    private final Object lifecycleLock = new Object();
    private final TaskWorkerAllocationPacer allocation;
    private final TaskRunningActivationPacer activation;
    private final TaskDispatchPacer dispatch;
    private Thread allocationThread;
    private Thread activationThread;
    private Thread dispatchThread;
    private CountDownLatch stopSignal;
    private State state = State.STOPPED;

    public AssignmentDispatchApplication(
            TaskWorkerAllocationPacer allocation,
            TaskRunningActivationPacer activation,
            TaskDispatchPacer dispatch
    ) {
        this.allocation = java.util.Objects.requireNonNull(
                allocation,
                "allocation"
        );
        this.activation = java.util.Objects.requireNonNull(
                activation,
                "activation"
        );
        this.dispatch = java.util.Objects.requireNonNull(dispatch, "dispatch");
    }

    public void start(AssignmentDispatchApplicationConfig config) {
        java.util.Objects.requireNonNull(config, "config");
        synchronized (lifecycleLock) {
            if (state != State.STOPPED
                    || allocationThread != null
                    || activationThread != null
                    || dispatchThread != null) {
                throw new IllegalStateException(
                        "Assignment Dispatch is already started"
                );
            }
            CountDownLatch signal = new CountDownLatch(1);
            allocationThread = loopThread(
                    "assignment-dispatch-worker-allocation",
                    signal,
                    config.workerAllocationIntervalMillis(),
                    () -> allocation.allocateCandidateWorkers(
                            config.workerAllocation()
                    )
            );
            activationThread = loopThread(
                    "assignment-dispatch-running-activation",
                    signal,
                    config.runningActivationIntervalMillis(),
                    () -> activation.activateRunningVisibleTasks(
                            config.runningActivation()
                    )
            );
            dispatchThread = loopThread(
                    "assignment-dispatch-task-dispatch",
                    signal,
                    config.taskDispatchIntervalMillis(),
                    () -> dispatch.dispatchTasks(config.taskDispatch())
            );
            stopSignal = signal;
            state = State.STARTING;
            List<Thread> started = new ArrayList<>(3);
            try {
                allocationThread.start();
                started.add(allocationThread);
                activationThread.start();
                started.add(activationThread);
                dispatchThread.start();
                started.add(dispatchThread);
                state = State.RUNNING;
            } catch (RuntimeException failure) {
                signal.countDown();
                started.forEach(Thread::interrupt);
                joinStarted(started, START_ROLLBACK_TIMEOUT_MILLIS, failure);
                allocationThread = null;
                activationThread = null;
                dispatchThread = null;
                stopSignal = null;
                state = State.FAILED;
                throw failure;
            }
        }
    }

    public void stop(long timeoutMillis) {
        if (timeoutMillis <= 0) {
            throw new IllegalArgumentException(
                    "timeoutMillis must be positive"
            );
        }
        List<Thread> threads;
        CountDownLatch signal;
        synchronized (lifecycleLock) {
            if (stopSignal == null) {
                return;
            }
            state = State.STOPPING;
            signal = stopSignal;
            threads = List.of(
                    allocationThread,
                    activationThread,
                    dispatchThread
            );
            signal.countDown();
        }
        long deadline = System.nanoTime()
                + TimeUnit.MILLISECONDS.toNanos(timeoutMillis);
        for (Thread thread : threads) {
            long remaining = Math.max(1, deadline - System.nanoTime());
            try {
                thread.join(Math.max(
                        1,
                        TimeUnit.NANOSECONDS.toMillis(remaining)
                ));
            } catch (InterruptedException error) {
                Thread.currentThread().interrupt();
                throw new IllegalStateException(
                        "Assignment Dispatch shutdown was interrupted",
                        error
                );
            }
        }
        if (threads.stream().anyMatch(Thread::isAlive)) {
            throw new IllegalStateException(
                    "Assignment Dispatch did not stop within its budget"
            );
        }
        synchronized (lifecycleLock) {
            allocationThread = null;
            activationThread = null;
            dispatchThread = null;
            stopSignal = null;
            state = State.STOPPED;
        }
    }

    public boolean isRunning() {
        synchronized (lifecycleLock) {
            refreshDeadThread();
            return state == State.RUNNING;
        }
    }

    public String state() {
        synchronized (lifecycleLock) {
            refreshDeadThread();
            return state.name();
        }
    }

    private Thread loopThread(
            String name,
            CountDownLatch signal,
            long intervalMillis,
            Runnable round
    ) {
        Thread thread = new Thread(
                () -> runLoop(name, signal, intervalMillis, round),
                name
        );
        thread.setDaemon(false);
        return thread;
    }

    private void runLoop(
            String operation,
            CountDownLatch signal,
            long intervalMillis,
            Runnable round
    ) {
        try {
            while (signal.getCount() > 0) {
                try {
                    round.run();
                } catch (RuntimeException error) {
                    LOGGER.log(
                            System.Logger.Level.ERROR,
                            "operation=" + operation + ".round failed",
                            error
                    );
                }
                if (signal.await(intervalMillis, TimeUnit.MILLISECONDS)) {
                    return;
                }
            }
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
        } finally {
            synchronized (lifecycleLock) {
                if (state == State.RUNNING) {
                    state = State.FAILED;
                    signal.countDown();
                }
            }
        }
    }

    private void refreshDeadThread() {
        if (state == State.RUNNING
                && (allocationThread == null || !allocationThread.isAlive()
                || activationThread == null || !activationThread.isAlive()
                || dispatchThread == null || !dispatchThread.isAlive())) {
            state = State.FAILED;
            if (stopSignal != null) {
                stopSignal.countDown();
            }
        }
    }

    private static void joinStarted(
            List<Thread> threads,
            long timeoutMillis,
            RuntimeException startFailure
    ) {
        long deadline = System.nanoTime()
                + TimeUnit.MILLISECONDS.toNanos(timeoutMillis);
        for (Thread thread : threads) {
            try {
                thread.join(Math.max(
                        1,
                        TimeUnit.NANOSECONDS.toMillis(Math.max(
                                1,
                                deadline - System.nanoTime()
                        ))
                ));
            } catch (InterruptedException error) {
                Thread.currentThread().interrupt();
                startFailure.addSuppressed(error);
                return;
            }
        }
        threads.stream().filter(Thread::isAlive).forEach(thread ->
                startFailure.addSuppressed(new IllegalStateException(
                        "Assignment Dispatch start rollback left thread "
                                + thread.getName() + " alive"
                ))
        );
    }

    private enum State {
        STOPPED,
        STARTING,
        RUNNING,
        STOPPING,
        FAILED
    }
}
